### PR TITLE
Evaluator

### PR DESCRIPTION
**Pull request by amjltc295**
## Why do we need this PR?
* To add evaluator #9 

## How is it implemented?
* Add EvaluatationPipeline and Evaluator
  * Evaluator may need to load results from other models, so a result data loader needs to be given. See `data_loaders/mnist_result.py` for example
  * EvaluatationPipeline and Evaluator do not call there parents' initialization
* Several small refactor
* Separate train/test/eval template configs

## Other changes
* Fix tester shuffled setting

### CODE REVIEW

File: src/configs/template.json [hunk 1]
```diff
@@ -76,5 +74,8 @@
             }
         }
     },
+    "log_step": 500,
+    "verbosity": 2,
+    "saved_keys": ["data_input", "data_target", "model_output"],
```
Comment (Nash2325138) [inline src/configs/template.json]: These keys are specifically for testing, which should be put in somewhere like data/mnist_test.json

File: src/configs/template_test_config.json [hunk 2]
```diff
@@ -0,0 +1,50 @@
+{
```
Comment (Nash2325138) [inline src/configs/template_test_config.json]: This file should be named as `template_evaluation_config.json`

File: src/data_loader/mnist_result.py [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+
+class MnistResultDataset(Dataset):
+    """
+    Customized MNIST result dataset demo
+    """
+    def __init__(self, result_filename):
+        self.results = self._load_data(result_filename)
+
+    def _load_data(self, result_filename):
+        return np.load(result_filename)['model_output']
```
Comment (Nash2325138) [inline src/data_loader/mnist_result.py]: Key `model_output` should be configurable since the model might change its output key anytime

File: src/pipeline/evaluation_pipeline.py [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+        workers = []
+        # Add a tester for each data loader
+        for gt_data_loader, result_data_loader in zip(self.gt_data_loaders, self.result_data_loaders):
+            evaluator = Evaluator(self, gt_data_loader, result_data_loader)
+            workers += [evaluator]
+        return workers
+
+    def run(self):
+        """
+        Full testing pipeline logic
```
Comment (Nash2325138) [inline src/pipeline/evaluation_pipeline.py]: `testing` -> `evaluation`

File: src/worker/evaluator.py [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+        return {'log': log}
+
+    def _to_log(self, epoch_stats):
+        return {}
+
+    def _print_log(self, epoch, batch_idx, batch_start_time, loss, metrics):
+        logger.info(
+            f'Epoch: {epoch} [{batch_idx * self.gt_data_loader.batch_size}/{self.gt_data_loader.n_samples} '
+            f' ({100.0 * batch_idx / len(self.gt_data_loader):.0f}%)] '
+            f'BT: {time.time() - batch_start_time:.2f}s'
```
Comment (Nash2325138) [inline src/worker/evaluator.py]: Using new variables for things like `batch_idx * self.gt_data_loader.batch_size` or `{100.0 * batch_idx / len(self.gt_data_loader)` will be more readible.
The same issues exist in other workers.

File: src/worker/worker_template.py [hunk 6]
```diff
@@ -97,7 +94,10 @@ def _get_and_write_metrics(self, data, model_output):
         acc_metrics = np.zeros(len(self.evaluation_metrics))
         for i, metric in enumerate(self.evaluation_metrics):
             acc_metrics[i] += metric(data, model_output)
-            self.writer.add_scalar(metric.nickname, acc_metrics[i])
+            try:
+                self.writer.add_scalar(metric.nickname, acc_metrics[i])
+            except AttributeError:
+                pass
```
Comment (Nash2325138) [inline src/worker/worker_template.py]: When will `AttributeError` be raised?

File: src/pipeline/base_pipeline.py [hunk 7]
```diff
@@ -46,10 +46,17 @@ def __init__(
         if args.pretrained is not None:
             self._load_pretrained(args.pretrained)
 
+        self.worker_outputs = {}
```
Comment (Nash2325138) [inline src/pipeline/base_pipeline.py]: Why do we need `worker_outputs` to be an attribute?

File: src/configs/template_test_config.json [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+                "nickname": "Top5Acc",
+                "k": 5,
+                "output_key": "model_output",
+                "target_key": "data_target"
+            }
+        }
+    },
+    "log_step": 500,
+    "verbosity": 2,
+    "name": "test_template_config"
```
Comment (Nash2325138) [inline src/configs/template_test_config.json]: Also this should be renamed.

File: src/pipeline/base_pipeline.py (hunk 7 above)
Comment (amjltc295) [reply src/pipeline/base_pipeline.py]: It has the potential to be passed to the following workers

File: src/configs/template.json (hunk 1 above)
Comment (amjltc295) [reply src/configs/template.json]: Fixed; train/test/eval configs are separated now

File: src/configs/template_test_config.json (hunk 2 above)
Comment (amjltc295) [reply src/configs/template_test_config.json]: Fixed

File: src/data_loader/mnist_result.py (hunk 3 above)
Comment (amjltc295) [reply src/data_loader/mnist_result.py]: Fixed

File: src/pipeline/evaluation_pipeline.py (hunk 4 above)
Comment (amjltc295) [reply src/pipeline/evaluation_pipeline.py]: Fixed

File: src/worker/evaluator.py (hunk 5 above)
Comment (amjltc295) [reply src/worker/evaluator.py]: Fixed

File: src/worker/worker_template.py (hunk 6 above)
Comment (amjltc295) [reply src/worker/worker_template.py]: See the comments

File: README.md [hunk 9]
```diff
@@ -19,12 +19,15 @@ python main.py --resume ./saved/ckpts/template_config+CrossEntropy/0723_180600/c
...
-python main.py -sc configs/data/mnist_test.json --mode test -p <pretrained_weight>
+# For submission to the leaderboard, etc.
+python main.py --mode test -p <pretrained_weight>
 ```
 
 Example: evaluate results
 ```
-python main.py --mode eval -tc configs/template_test_config.json
+# You could evaluate results by other models (corresponding data loader needs to be defined)
+# See configs/template_eval_config.json for details
```
Comment (Nash2325138) [inline README.md]: Emphasizing entry `result_data_loaders` in this comment will be better.

File: src/data_loader/data_loaders.py [hunk 10]
```diff
@@ -31,8 +31,9 @@ class MnistResultDataLoader(BaseDataLoader):
     Customized MNIST result data loader demo
     Returned data will be in dictionary
     """
-    def __init__(self, result_filename, batch_size, num_workers=1, training=True, name=None):
+    def __init__(self, result_filename, key, batch_size, num_workers=1, training=True, name=None):
         self.result_filename = result_filename
```
Comment (Nash2325138) [inline src/data_loader/data_loaders.py]: Should `result_filename` be an attribute of `MnistResultDataLoader`?
Or is it specific for `MnistResultDataset`?

Comment (Nash2325138) [reply src/data_loader/data_loaders.py]: If it is, passing `result_filename` and `key` as a dictionary via the config file is a better way for the compliance of the open-closed principle.

File: src/worker/worker_template.py (hunk 6 above)
Comment (Nash2325138) [reply src/worker/worker_template.py]: Decouple `_get_metrics()` and `_write_metrics()` and overwrite _write_metrics in the evaliuator would be a better way, imo.
